ompi/mca/mtl/psm2/mtl_psm2_endpoint.h

```diff
@@ -58,7 +58,13 @@ OBJ_CLASS_DECLARATION(mca_mtl_psm2_endpoint);
 static inline mca_mtl_psm2_endpoint_t *ompi_mtl_psm2_get_endpoint (struct mca_mtl_base_module_t* mtl, ompi_proc_t *ompi_proc)
 {
     if (OPAL_UNLIKELY(NULL == ompi_proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_MTL])) {
-	ompi_mtl_psm2_add_procs (mtl, 1, &ompi_proc);
+        if (OPAL_UNLIKELY(OMPI_SUCCESS != MCA_PML_CALL(add_procs(&ompi_proc, 1)))) {
+            /* Fatal error. exit() out */
+            opal_output(0, "%s:%d: *** The Open MPI PSM2 MTL is aborting the MPI job (via exit(3)).\n",
+                           __FILE__, __LINE__);
+            fflush(stderr);
+            exit(1);
```

Comment:
Forgot to mention: I don't think you're supposed to call `exit()` directly; aren't you supposed to call `module->pml_error_callback` (assuming `module` and `module->pml_error_callback` are non-NULL)?

Reply:
Same as my previous comment, this matches the code in #7571 .

Reply:
I don't see anything like that in the referenced PR - was it previously in that code area?

Reply:
Look at the commit for #7571, or lines 45-47 of mtl_ofi_endpoint.h. This is literally a cut-and-paste of Brian's code.

Reply:
(well, brian's code plus existing code in that function.)

Reply:
That is what I was asking, Mike - when you say this is something previously reviewed and approved, we only looked at the code that was actually being changed. Nobody looked at the rest of it or we would have flagged it there. Libraries are _not_ allowed to call `exit` under any circumstances. The `fflush` call is likely some old cruft.

Reply:
Well, like I said, I'm willing to change it.

So, the previous code always just assumed add_procs() couldn't fail. Looking at other MTLs, OFI calls exit() (as we've discussed) and portals4 doesn't call add_procs(). Looking at the OFI and PSM2 MTLs, both of them will crash if add_procs() fails. 

I searched the code for references to pml_error_callback and the only place it's used is in usenic.  There are, however, a number of places in portals4 and osc/pt2pt that call abort() directly and dump core.

I mean, I could call ompi_mpi_abort()?